hermes/src/network/rpc/ws.rs

```diff
@@ -0,0 +1,290 @@
+use {
+    super::types::{
+        PriceIdInput,
+        RpcPriceFeed,
+    },
+    crate::store::Store,
+    anyhow::Result,
+    axum::{
+        extract::{
+            ws::{
+                Message,
+                WebSocket,
+                WebSocketUpgrade,
+            },
+            State,
+        },
+        response::IntoResponse,
+    },
+    dashmap::DashMap,
+    futures::{
+        stream::{
+            SplitSink,
+            SplitStream,
+        },
+        SinkExt,
+        StreamExt,
+    },
+    pyth_sdk::PriceIdentifier,
+    serde::{
+        Deserialize,
+        Serialize,
+    },
+    std::{
+        collections::HashMap,
+        sync::atomic::{
+            AtomicUsize,
+            Ordering,
+        },
+    },
+    tokio::sync::mpsc,
+};
+
+
+pub async fn ws_route_handler(
+    ws: WebSocketUpgrade,
+    State(state): State<super::State>,
+) -> impl IntoResponse {
+    ws.on_upgrade(|socket| websocket_handler(socket, state))
+}
+
+async fn websocket_handler(stream: WebSocket, state: super::State) {
+    let ws_state = state.ws.clone();
+    let id = ws_state.subscriber_counter.fetch_add(1, Ordering::SeqCst);
+
+    let (sender, receiver) = stream.split();
+
+    // TODO: Use a configured value for the buffer size or make it const static
+    // TODO: Use redis stream to source the updates instead of a channel
+    let (tx, rx) = mpsc::channel::<Vec<PriceIdentifier>>(1000);
+
+    ws_state.subscribers.insert(id, tx);
+
+    log::debug!("New websocket connection, assigning id: {}", id);
+
+    let mut subscriber = Subscriber::new(id, state.store.clone(), rx, receiver, sender);
+
+    subscriber.run().await;
+}
+
+pub type SubscriberId = usize;
+
+/// Subscriber is an actor that handles a single websocket connection.
+/// It listens to the store for updates and sends them to the client.
+pub struct Subscriber {
+    id:                      SubscriberId,
+    closed:                  bool,
+    store:                   Store,
```

Comment:
This kind of dependency is not great. We should rethink about the structure in the coming refactors.

Reply:
There are locks to prevent misuse (Rust doesn't allow otherwise :D) but I also created the direction of dependency in a way that it doesn't introduce a deadlock.

Reply:
I think generally it's fine. I don't like it much because it's half actor and half part of a global component.

I could store the Subscriber as a global component and let the global dispatcher directly call the methods and I tried. The result wasn't great because the Subscriber has a stream coming from the client and when it gets closed on destruction it should remove itself from the global reference to it (which it has access to the reference) and ...

I think the actor model offers better decoupling but that's a bigger design decision that I didn't want to make on this PR.